tensorflow/python/ops/array_ops.py

```diff
@@ -2085,7 +2085,13 @@ def _normalize_sparse_shape(shape, name):
     return (None, None)
   rank = shape.get_shape()[0] if isinstance(shape, ops.Tensor) else len(shape)
   if not isinstance(shape, ops.Tensor) and None in shape:
-    return (None, rank)
+    if not isinstance(shape, (list, tuple)):
+      return (None, rank)
+    # In case shape is a list or tuple, we will convert None into
+    # -1 (Unknwon dims) so that the shape is preserved when calling
+    # convert_to_tensor below. This will make sure tf.sparse.placeholder
+    # accepts `(None, 4)` like tf.placeholder.
+    shape = [dim if dim is not None else -1 for dim in shape]
```

Comment:
I don't think putting -1s here is the right fix. Specifically, I worry that downstream users of that shape tensor will not be able to process the -1s in the elements.

My understanding of the shape tensor in sparse tensors is that, when evaluated, it should be the full shape of the sparse tensor, which this change breaks.

Can you add a longer test for the kind of code you want to support?